### PR TITLE
Allow `@CheckForNull` annotations to override `@ParametersAreNonnullByDefault` in DataBoundContructor

### LINKED ISSUE #1025: DataBoundConfigurator enforces Constructor fields when ParametersAreNonnullByDefault is set, regardless of the CheckForNull flag
It is a follow-up to the discussion in https://github.com/SonarSource/sonar-scanner-jenkins/pull/124#discussion_r317989554 with @Smasherr . https://github.com/jenkinsci/configuration-as-code-plugin/blob/f72058445c0250041303a73b5d8566140d337c67/plugin/src/main/java/io/jenkins/plugins/casc/impl/configurators/DataBoundConfigurator.java#L109-L137 checks `ParametersAreNonnullByDefault ` and fails for a null para,eter, even if a `CheckForNull` or `Nullable` annotation is set for a constructor parameter.

AFAICT it is a regression in https://github.com/jenkinsci/configuration-as-code-plugin/releases/tag/configuration-as-code-1.24 after the annotation handling enhancements in https://github.com/jenkinsci/configuration-as-code-plugin/pull/950. CC @AbhyudayaSharma

### PR DESCRIPTION
**Pull request by timja**
Previously if someone had put ParametersAreNonnullByDefault on a class /
method / package-info there was no way to override and say this field
should be nullable.

Fixes #1025

### Your checklist for this pull request

🚨 Please review the [guidelines for contributing](../blob/master/docs/CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Did you provide a test-case? That demonstrates feature works or fixes the issue.

### CODE REVIEW

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/configurators/DataBoundConfigurator.java [hunk 1]
```diff
@@ -128,8 +129,6 @@ private T tryConstructor(Constructor<T> constructor, Mapping config, Configurati
                         LOGGER.log(Level.FINER, "The parameter to be set is @Nonnull but is not present; " +
                                                            "setting equal to empty list.");
                         args[i] = Collections.emptyList();
-                    } else if (String.class.isAssignableFrom(t)) {
```
Comment (timja) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/configurators/DataBoundConfigurator.java]: this is a behaviour change but as far as I can tell this should never have been merged,
It was a hack to make sonar plugin work better, but sonar plugin should just be fixed.

https://github.com/jenkinsci/configuration-as-code-plugin/pull/950/files#r304891125

**Comment by oleg-nenashev** [comment]
Should it also consider other CheckForNull implementations?

**Comment by timja** [comment]
> Should it also consider other CheckForNull implementations?

Findbugs isn't retained for reflective access

**Comment by timja** [comment]
> Should it also consider other CheckForNull implementations?

Do you mean Nullable? I can add javax.annotation Nullable but not findbugs as it has the same issue with retention

Or we don't implement this and say Constructor parameters are mandatory and databoundsetters are for optional fields...

**Comment by casz** [comment]
CheckForNull and co is actually covered by javax.annotations

https://github.com/spotbugs/spotbugs/blob/master/spotbugs-annotations/src/main/java/edu/umd/cs/findbugs/annotations/CheckForNull.java

Problem with Spotbugs annotation is that have an retention policy

**Comment by timja** [comment]
> > Do you mean Nullable? I can add javax.annotation Nullable but not findbugs as it has the same issue with retention
> 
> I mean the fact we ignore `edu.umd.cs.findbugs.annotations.CheckForNull` & Co. It is a step forward anyway, but it would be nice to cover other common annotation types. I do not care much about `Nullable`, because it is not a best practie for static analysis anyway.
> 

Can't use them due to the retention policy

> P.S: bonus points for better changelog entry

Done

**Comment by oleg-nenashev** [comment]
> Problem with Spotbugs annotation is that have an retention policy

Something we can figure out later.
Maybe including SpotBugs annotation processing libraries in JCasC is not a bad idea after all...

**Comment by dcoraboeuf** [comment]
Hi, do you know when this PR will be released?

**Comment by timja** [comment]
I've just asked in gitter about it, https://gitter.im/jenkinsci/configuration-as-code-plugin

There's an incremental version here if it's something you're developing against:
https://repo.jenkins-ci.org/incrementals/io/jenkins/configuration-as-code/1.31-rc915.effa048b2e97/

File: plugin/src/test/java/io/jenkins/plugins/casc/impl/configurators/nonnull/nonnullparampackage/PackageParametersNonNullCheckForNull.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+/**
+ * Checks the behaviour of {@link io.jenkins.plugins.casc.impl.configurators.DataBoundConfigurator} with non null
+ * {@link Secret} when using package-level {@link javax.annotation.ParametersAreNonnullByDefault} annotations.
+ */
+public class PackageParametersNonNullCheckForNull {
+    private Secret secret;
+
+    @DataBoundConstructor
+    public PackageParametersNonNullCheckForNull(@CheckForNull Secret secret) {
```
[the remark this answers is not in the document]
Comment (timja) [reply plugin/src/test/java/io/jenkins/plugins/casc/impl/configurators/nonnull/nonnullparampackage/PackageParametersNonNullCheckForNull.java]: Yes jcasc does this by default This general feature predates my maintainer ship here. I assume it was to reduce changes required in plugins to be made compatible / give them a less effort way with lower backwards compatibility risk.

**Pull request merged**